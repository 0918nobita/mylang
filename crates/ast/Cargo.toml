[package]
name = "mylang_ast"
version = "0.1.0"
edition = "2021"

[dependencies]
mylang_token = { path = "../token" }
serde = { version = "1.0.167", features = ["derive"] }
