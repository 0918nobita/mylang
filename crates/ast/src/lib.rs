mod expr;
mod stmt;

pub use expr::Expr;
pub use stmt::Stmt;
