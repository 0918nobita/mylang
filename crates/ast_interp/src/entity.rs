//! 抽象構文木インタプリタで扱う「実体」の定義

use std::fmt::{Debug, Display, Formatter, Result as FmtResult};

/// 符号付き32ビット整数値
pub struct I32Entity(i32);

impl I32Entity {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn add(&self, rhs: &Self) -> Self {
        Self(self.0 + rhs.0)
    }
}

impl Debug for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// 文字列データ
pub struct StrEntity(String);

impl StrEntity {
    pub fn new(value: String) -> Self {
        Self(value)
    }
}

impl Debug for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// 実行時型情報
#[derive(Debug)]
pub enum RuntimeTypeInfo {
    I32,
    Str,
}

impl Display for RuntimeTypeInfo {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(
            f,
            "{}",
            match self {
                RuntimeTypeInfo::I32 => "i32",
                RuntimeTypeInfo::Str => "str",
            }
        )
    }
}

/// バイトコード実行時に扱う値
#[derive(Debug)]
pub enum Entity {
    I32(I32Entity),
    Str(StrEntity),
}

impl Entity {
    pub fn get_type(&self) -> RuntimeTypeInfo {
        match self {
            Entity::I32(_) => RuntimeTypeInfo::I32,
            Entity::Str(_) => RuntimeTypeInfo::Str,
        }
    }
}
