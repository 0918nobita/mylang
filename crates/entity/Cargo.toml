[package]
name = "mylang_entity"
version = "0.1.0"
edition = "2021"

[dependencies]
