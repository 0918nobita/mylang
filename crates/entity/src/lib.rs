use std::fmt::{Debug, Formatter, Result as FmtResult};

use serde::Serialize;

#[derive(PartialEq, Serialize)]
pub struct I32Entity(i32);

impl I32Entity {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn add(&self, rhs: &Self) -> Self {
        Self(self.0 + rhs.0)
    }
}

impl Debug for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

#[derive(PartialEq, Serialize)]
pub struct StrEntity(String);

impl StrEntity {
    pub fn new(value: String) -> Self {
        Self(value)
    }
}

impl Debug for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

#[derive(Debug, Serialize)]
pub enum Entity {
    I32(I32Entity),
    Str(StrEntity),
}
