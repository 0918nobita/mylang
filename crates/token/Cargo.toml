[package]
name = "mylang_token"
version = "0.1.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.143", features = ["derive"] }
