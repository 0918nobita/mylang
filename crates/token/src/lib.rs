//! 字句解析で取り出されるトークンとその付帯情報の定義

mod locatable;
mod pos;
mod range;

use serde::{Deserialize, Serialize};

pub use locatable::Locatable;
pub use pos::Pos;
pub use range::Range;

/// [`Pos`] を簡単に生成するためのマクロ
#[macro_export]
macro_rules! pos {
    ($line:expr ; $char:expr) => {
        mylang_token::Pos::new($line, $char)
    };
}

/// [`Range`] を簡単に生成するためのマクロ
#[macro_export]
macro_rules! range {
    ($start_line:expr ; $start_char:expr, $end_line:expr ; $end_char:expr) => {
        mylang_token::Range::new(
            mylang_token::pos!($start_line;$start_char),
            mylang_token::pos!($end_line;$end_char),
        )
    };
    ($start:expr, $end:expr) => {
        mylang_token::Range::new($start, $end)
    };
}

#[derive(Debug, Serialize, Deserialize)]
pub enum KeywordKind {
    Let,
    In,
    PrintI32,
    PrintStr,
}

impl KeywordKind {
    pub fn parse(str: &str) -> Option<Self> {
        match str {
            "let" => Some(Self::Let),
            "in" => Some(Self::In),
            "print_int" => Some(Self::PrintI32),
            "print_str" => Some(Self::PrintStr),
            _ => None,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Token {
    I32(Range, i32),
    AddOp(Pos),
    Str(Range, String),
    Keyword(Range, KeywordKind),
    Ident(Range, String),
    Newline(Pos),
}

impl Locatable for Token {
    fn locate(&self) -> Range {
        match self {
            Self::I32(r, _) | Self::Str(r, _) | Self::Keyword(r, _) | Self::Ident(r, _) => {
                r.clone()
            }

            Self::AddOp(p) | Self::Newline(p) => p.clone().into(),
        }
    }
}
