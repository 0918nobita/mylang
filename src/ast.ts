export type NullLiteral = Readonly<{
    type: 'nullLiteral';
}>;

export type BooleanLiteral = Readonly<{
    type: 'booleanLiteral';
    value: boolean;
}>;

export type NumberLiteral = Readonly<{
    type: 'numberLiteral';
    value: number;
}>;

export type StringLiteral = Readonly<{
    type: 'stringLiteral';
    value: string;
}>;

export type Literal =
    | NullLiteral
    | BooleanLiteral
    | NumberLiteral
    | StringLiteral;

export type Identifier = Readonly<{
    type: 'identifier';
    name: string;
}>;

export type BinaryExpression = Readonly<{
    type: 'binaryExpression';
    operator: '+' | '*';
    left: Expression;
    right: Expression;
}>;

export type CallExpression = Readonly<{
    type: 'callExpression';
    callee: Expression;
    arguments: Expression[];
    optional: boolean;
}>;

export type MemberExpression = Readonly<{
    type: 'memberExpression';
    object: Expression;
    property: Expression;
    computed: boolean;
    optional: boolean;
}>;

export type Expression =
    | Literal
    | Identifier
    | BinaryExpression
    | CallExpression
    | MemberExpression;
