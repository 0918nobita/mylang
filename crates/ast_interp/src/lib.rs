use anyhow::bail;
use ast::{Expr, Stmt};
use entity::{Entity, I32Entity, StrEntity};

fn eval(expr: &Expr) -> anyhow::Result<Entity> {
    match expr {
        Expr::I32Lit(_, i) => Ok(Entity::I32(I32Entity::new(*i))),

        Expr::StrLit(_, s) => Ok(Entity::Str(StrEntity::new(s.clone()))),

        Expr::Add(ref lhs, ref rhs) => {
            let lhs = eval(lhs)?;
            let rhs = eval(rhs)?;

            if let (Entity::I32(lhs), Entity::I32(rhs)) = (lhs, rhs) {
                Ok(Entity::I32(lhs.add(&rhs)))
            } else {
                bail!("Type mismatch")
            }
        }
    }
}

pub fn execute(stmts: &[Stmt]) -> anyhow::Result<()> {
    for stmt in stmts {
        match stmt {
            Stmt::PrintI32(_, ref expr) => {
                let val = eval(expr)?;
                if let Entity::I32(ent) = val {
                    println!("{:?}", ent);
                } else {
                    bail!("Type mismatch");
                }
            }
            Stmt::PrintStr(_, ref expr) => {
                let val = eval(expr)?;
                if let Entity::Str(ent) = val {
                    println!("{:?}", ent);
                } else {
                    bail!("Type mismatch");
                }
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use ast::Expr;
    use token::Range;

    use super::eval;

    #[test]
    fn test_i32_lit() {
        let ent = eval(&Expr::I32Lit(Range::default(), 2));
        insta::assert_debug_snapshot!(ent);
    }

    #[test]
    fn test_i32_lit_plus_i32_lit() {
        let ent = eval(&Expr::Add(
            Box::new(Expr::I32Lit(Range::default(), 2)),
            Box::new(Expr::I32Lit(Range::default(), 3)),
        ));
        insta::assert_debug_snapshot!(ent);
    }

    #[test]
    fn test_str_lit() {
        let ent = eval(&Expr::StrLit(Range::default(), "foo".to_owned()));
        insta::assert_debug_snapshot!(ent);
    }

    #[test]
    fn test_i32_lit_plus_str_lit() {
        let res = eval(&Expr::Add(
            Box::new(Expr::I32Lit(Range::default(), 3)),
            Box::new(Expr::StrLit(Range::default(), "bar".to_owned())),
        ));
        insta::assert_debug_snapshot!(res);
    }

    #[test]
    fn test_str_lit_plus_i32_lit() {
        let res = eval(&Expr::Add(
            Box::new(Expr::StrLit(Range::default(), "foo".to_owned())),
            Box::new(Expr::I32Lit(Range::default(), 4)),
        ));
        insta::assert_debug_snapshot!(res);
    }

    #[test]
    fn test_str_lit_plus_str_lit() {
        let res = eval(&Expr::Add(
            Box::new(Expr::StrLit(Range::default(), "foo".to_owned())),
            Box::new(Expr::StrLit(Range::default(), "bar".to_owned())),
        ));
        insta::assert_debug_snapshot!(res);
    }
}
