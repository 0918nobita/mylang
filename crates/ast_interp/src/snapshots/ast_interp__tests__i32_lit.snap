---
source: crates/ast_interp/src/lib.rs
assertion_line: 59
expression: ent

---
Ok(
    I32(
        2,
    ),
)
